### DISCUSSION TITLE
Question regarding the implementation of the bluetooth driver on technic hub

### DISCUSSION THREAD

**Discussion by NStrijbosch** [pybricks/support#416] [Q&A]
I am exploring the possibility to connect a LEGO Mario to a Technic Hub running pybricks. 

At this point I have established a stable connection! 
But to achieve this I had to comment out a small piece of code:

https://github.com/pybricks/pybricks-micropython/blob/master/lib/pbio/drv/bluetooth/bluetooth_stm32_cc2640.c#:~:text=//%20Assuming%20that%20we,%7D)%3B

I do not understand the purpose of these few lines, and I am just wondering what the impact of removing these lines might be. 
Moreover, any ideas why these lines are blocking the connection process of the Mario but are not blocking when connecting the Remote or City hub?

**Comment by dlech** [comment]
The link just shows the entire file. It might be easier to just push your changes to GitHub so we can see what you changed.

**Reply by NStrijbosch** [reply]
Hm, the link worked for me. Anyhow it are lines 499-506 from the specific file. 

Let me see how I can push the necessary changes

**Reply by dlech** [reply]
OK, so this section:

```C
    // Assuming that we only ever get successful ATT_ReadByTypeRsp and failed
    // ATT_ReadByTypeRsp or ATT_ErrorRsp.
    PT_WAIT_UNTIL(pt, {
        if (task->cancel) {
            goto cancel_disconnect;
        }
        remote_lwp3_char_handle != NO_CONNECTION;
    });
```

**Reply by NStrijbosch** [reply]
I believe I pushed the adjustments. (Note that the Remote won't work anymore)

**Reply by NStrijbosch** [reply]
> OK, so this section:
> 
> ```c
>     // Assuming that we only ever get successful ATT_ReadByTypeRsp and failed
>     // ATT_ReadByTypeRsp or ATT_ErrorRsp.
>     PT_WAIT_UNTIL(pt, {
>         if (task->cancel) {
>             goto cancel_disconnect;
>         }
>         remote_lwp3_char_handle != NO_CONNECTION;
>     });
> ```

Exactly!

**Reply by dlech** [reply]
`remote_lwp3_char_handle` has to be set to a valid value eventually, otherwise the connection could not be used. So I'm not sure why this would be blocking.

**Reply by NStrijbosch** [reply]
Clear, so that means the connection to Mario is still not complete if I comment these lines. 
And that answers the question: what could go wrong? -> communication is not possible?

**Reply by dlech** [reply]
That value is used in the next command to enable notifications on the LWP3 Hub characteristic. so it if was still set to `NO_CONNECTION`, then notifications would not be enabled and no data would be received from Mario.

**Reply by dlech** [reply]
To help figure out what is going on, it would be useful to connect Mario to a computer (e.g. using `pybricksdev lwp3 repl`) and record the Bluetooth packets using Wireshark while doing this. This will let us see what the Mario hub is actually sending over the air.

**Reply by NStrijbosch** [reply]
While installing wireshark I did some test with the remote. Commenting out those lines indeed prevents communication... 

I guess I was a little bit to happy when my Mario didn't disconnect immediately.

**Reply by NStrijbosch** [reply]
[wireshark_connect_mario.csv](https://github.com/pybricks/support/files/6925259/wireshark_connect_mario.csv)

Attached some logs of the bluetooth communication using wireshark while connecting to Mario.

**Reply by dlech** [reply]
Great. I'm guessing Mario must not properly implement the command that `GATT_DiscCharsByUUID()` sends. This command is supposed to find the handle number for the LWP3 Hub characteristic. When connecting from the computer, it enumerates all characteristics instead of a specific one, so it uses a different command. As a hack we could try adding `remote_lwp3_char_handle = 0x0011;` in place of the code you commented out.

**Reply by NStrijbosch** [reply]
Yes, great.That did the trick!
Thanks for your quick response

**Reply by dlech** [reply]
FYI, I just pushed a couple more commits that you might find interesting.

**Reply by NStrijbosch** [reply]
Great! I was already expecting the last one ;)
